arch-design-questions.md

```diff
@@ -0,0 +1,93 @@
+
+# Single use clients
+
+Clients are single-use - you cannot change their receive mode or targeted queue/topic/subscription after
+they are created.
+
+This also means that calling .receive() (or registering multiple handlers) isn't a useful activity - the user
+would only be competing against _themselves_ since calling receive() multiple times would just result
+in multiple links, in the same mode, to the same entity, on the same _connection_.
+
+# Sessions design
+
+The primary use case for non-session bus users when receiving is opening
+up a single client instance, connecting to a targeted queue (or topic/subscription)
+and receiving from it.
+
+In this case the lifetime of the client is essentially the lifetime of their application which makes
+connection management a non-issue.
+
+Sessions differ in that the user's application is not tied to a single session - the user
+is expected to open up and read from many sessions during the lifetime of their application.
+
+This requires a few design considerations:
+
+## Connection "explosion"
+
+**Problem**
+
+The user opening multiple instances of a client means we need a way to limit the number of
+physical connections to service bus.
+
+Opening multiple receive _links_ is fine with service bus (it's a cheap abstraction within
+the physical connection itself). However, a user creating a standalone client instance (without
+any form of sharing) will inadvertantly create many physical connections.
+
+**Solution**
+
+Create a higher-lifetime object that the user can hold onto that represents the connection.
+This higher-lifetime object **must be required** when the user is working with sessions
+due to the higher likelihood of failure (via connection explosion).
+
+**Existing solutions:**
+
+- Java (keeping the service bus client builder alive + sharedConnection(true). Not required for sessions?
+- .NET (create ServiceBusConnection class, share. Not required though but should be for sessions.)
+- JS - { sessionConnections: } property when client is initialized (required if specifying a session)
+- Python - top level service bus client instance?
+
+## The user doesn't necessarily know which sessions exist
+
+**Problem**
+
+One use case with sessions is that the user doesn't necessarily know what sessions
+exist ahead of time.
+
+An anticipated common pattern is the user will do something like this:
+
+```
+  reusable-connection = create service bus connection
+
+  while:
+    // concurrently...
+    get next session(reusable-connection)
+    read from session until all message exhausted
+    close session
+```
+
+**get next session** can be handled by simply allowing the [single use client](#single-use-clients)
+to take a sentinel value for the session ID, allowing the server to just hand us the next unlocked session.
+
+We anticipate that this will be a fairly common pattern. To simplify the experience for the user each
+language will have some form of a **session manager**..
```

Comment:
This would correspond to the SessionClient in Track 1 .NET. I was hoping to avoid adding something like this particularly since it isn't needed for the more common event handling use case. For the pull based receive, users can pass in an existing connection to a new client instance to start pulling from a different session. Do we think that this is not sufficient to cover this use case?

Reply:
I _think_ that's what you've coded into the Processor class, correct?

Reply:
For the event based usage, this is covered in the Processor. For the pull based approach users would need to create new clients and pass in the ServiceBusConnection.

Reply:
So to the question we had in the arch committee  the other day - do we consider 'pull' to be "build this logic yourself" and thus we don't need to provide SessionManagerPullEdition?

Reply:
I would vote for that.

Reply:
@hemanttanwar , @conniey, @yunhaoling , @KieranBrantnerMagee , @ShivangiReja , @HarshaNalluru , @chradek  - what do you all think? Will you only offer this kind of manager functionality in one mode or another? (and if so - which?)

Or alternatively - are you _not_ planning on offering this?

Reply:
I would note the "manager" aspect of this is transparent to the user in .NET and doesn't involve additional types.

Reply:
Yes, @JoshLove-msft , from what I can tell you've basically got a single processor that handles both cases:
* Named session
* Unnamed multiple sessions

Reply:
This is what .NET calls "round-robin" sessions, right?   We should probably lock on a name for the concept, as I see it being one of those things the board comes back to for several discussions.

Reply:
"Round robin" resonates with everyone. I'll change it to be that instead.

Reply:
I added in round-robin session as the term and also updated the link section to point to the processor as the .NET implementation for this concept.